### PR TITLE
[module-scripts] Fix source-login-scripts.sh be generated in other expo modules

### PR DESCRIPTION
**Pull request by Kudo**
# Why

regression of #15336, where we should not sync `source-login-scripts.sh` to other expo modules.
i just misread the `if [ ! -f "$output_file" ] || grep --quiet "@generated" "$output_file"; then` line.

# How

- introduce `source-login-scripts.sh` as an optional template file which only to sync if it's present and contains `@generated` pattern as described in https://github.com/expo/expo/pull/15336#discussion_r758905459

# Test Plan

- `yarn prepare` in expo-device and make sure `source-login-scripts.sh` not be generated.
- change `source-login-scripts.sh` and `yarn prepare` in expo-constants. check it's updated.

# Checklist

- n/a Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
  - #15336 is not published yet, i'm not going to update changelog again.
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-module-scripts/bin/expo-module-configure [hunk 1]
```diff
@@ -6,8 +6,15 @@ script_dir="$(dirname "$0")"
...
 shopt -s dotglob
 
-# this function returns relative file paths inside a given directory.
-# usage: get_relative_files /path/to/dir
+# an optional template file will only be synced if target file is present in file system,
+# and its content contains `@generated` pattern.
+OPTIONAL_TEMPLATE_FILES=(
+  # add a relative file path from `templates/` for new optional file
+  "scripts/source-login-scripts.sh"
+)
```
Comment (ide) [inline packages/expo-module-scripts/bin/expo-module-configure:14]:
```suggestion
# an optional template file will be synced only if the target file is present in the file system,
# and its content contains the string `@generated`.
OPTIONAL_TEMPLATE_FILES=(
  # add a relative file path from `templates/` for each new optional file
  "scripts/source-login-scripts.sh"
)
```

File: packages/expo-module-scripts/bin/expo-module-configure [hunk 2]
```diff
@@ -6,8 +6,15 @@ script_dir="$(dirname "$0")"
...
+
+# returns relative file paths inside a given directory.
```
Comment (ide) [inline packages/expo-module-scripts/bin/expo-module-configure:16]:
```suggestion
# returns relative file paths inside a given directory without the leading "./".
```

File: packages/expo-module-scripts/bin/expo-module-configure [hunk 3]
```diff
@@ -16,13 +23,54 @@ get_relative_files() {
 }
 
 
+# sync file if target file is missing or the existing file contains `@generated` pattern.
```
Comment (ide) [inline packages/expo-module-scripts/bin/expo-module-configure:26]:
```suggestion
# syncs file if target file is missing or the existing file contains `@generated`.
```

File: packages/expo-module-scripts/bin/expo-module-configure [hunk 4]
```diff
@@ -16,13 +23,54 @@ get_relative_files() {
...
+  local source=$1
+  local target=$2
+  # echo "sync_file_if_present $source -> $target"
+  if [ -f "$target" ] && grep --quiet "@generated" "$target"; then
+    rsync --checksum "$source" "$target"
+  fi
+}
+
+# check if a file is listed in `OPTIONAL_TEMPLATE_FILES`.
+# usage: if is_optional_file "/path/to/file"; then ... fi
```
Comment (ide) [inline packages/expo-module-scripts/bin/expo-module-configure:49]: the optional files are all relative
```suggestion
# usage: if is_optional_file "path/to/file"; then ... fi
```

File: packages/expo-module-scripts/bin/expo-module-configure (hunk 3 above)
Comment (ide) [inline packages/expo-module-scripts/bin/expo-module-configure:26]:
```suggestion
# syncs the source file if the target file is missing or the existing file contains `@generated`.
```

File: packages/expo-module-scripts/bin/expo-module-configure [hunk 5]
```diff
@@ -16,13 +23,54 @@ get_relative_files() {
...
+sync_file_if_missing() {
+  local source=$1
+  local target=$2
+  # echo "sync_file_if_missing $source -> $target"
+  if [ ! -f "$target" ] || grep --quiet "@generated" "$target"; then
+    rsync --checksum "$source" "$target"
+  fi
+}
+
+# sync file if target file is present in file system and its content contains `@generated` pattern.
```
Comment (ide) [inline packages/expo-module-scripts/bin/expo-module-configure:37]:
```suggestion
# syncs the source file if the target file is present in the file system and its content contains `@generated`.
```

**Review by ide — APPROVED** [review]

**Comment by Kudo** [comment]
thanks for the nice suggestions and it's being clear. good to know the "An Xcode" 🤩

**Pull request merged**